Fix building c47edit using MinGW on Linux

madebr opened the pull request:
These patches fix building c47edit on Linux using a MinGW toolchain.

madebr committed:
Define NOMINMAX in vcxproj

madebr committed:
Fix Windows SDK headers for MinGW on Linux

madebr committed:
Convert c47edit.rc to utf-8 and remove BOM

madebr committed:
Add missing C++STL includes + include <filesystem>

madebr committed:
Fix 'dependent-name 'T::value_type' is parsed as a non-type, but instantiation yields a type'

madebr committed:
Fix error: explicit template specialization cannot have a storage class

madebr committed:
Fix error: cannot bind non-const lvalue reference to an rvalue

madebr committed:
Only build x64-windows-rel triplet on ci (no x64-windows-dbg)

madebr committed:
Fix error: explicit specialization in non-namespace scope

AdrienTD commented:
Thanks for the PR! It's indeed useful to support other compilers, and also make it compile on Linux (and run with Wine), because you never know how Windows will become these days...

But at the same time I would have to be more careful to not introduce some compile errors that would only trigger GCC/Clang but not VC++ :)

I will review it, there's some things I need to ask.

madebr commented on c47edit.vcxproj:
```diff
@@ -103,22 +103,30 @@
       <AdditionalDependencies>winmm.lib;opengl32.lib;glu32.lib;kernel32.lib;user32.lib;gdi32.lib;winspool.lib;comdlg32.lib;advapi32.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;odbc32.lib;odbccp32.lib;%(AdditionalDependencies)</AdditionalDependencies>
       <SubSystem>NotSet</SubSystem>
     </Link>
-    <ClCompile>
-      <LanguageStandard>stdcpp17</LanguageStandard>
-    </ClCompile>
```
I think I removed this accidentally.

madebr replied:
Never mind, I moved it up.

AdrienTD commented on ByteReader.h:
Did some search, it seems that specializations in non-namespace scope has actually been made allowed around C++17: https://cplusplus.github.io/CWG/issues/727.html
So not sure why you are getting an error there.

madebr replied:
I don't know why this occurs in gcc 13.2.
This is the error:
```
/home/maarten/projects/c47edit/PathfinderInfo.h:89:28: error: template argument 1 is invalid
   89 |         std::vector<uint8_t> toBytes() const;
      |                            ^
/home/maarten/projects/c47edit/PathfinderInfo.h:89:28: error: template argument 2 is invalid
In file included from /home/maarten/projects/c47edit/PathfinderInfo.cpp:2:
/home/maarten/projects/c47edit/ByteReader.h:25:18: error: explicit specialization in non-namespace scope 'class ByteReader'
   25 |         template<>
      |                  ^
```

C++ is a standard which compiler developers have to implement.
It's possible not everything is implemented..

AdrienTD replied:
Yeah, apparently [gcc does not support this yet](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85282), even though MSVC and Clang do.

A bit unfortunate, because now with the specialization outside the class, it is not next to the first template anymore, which is a bit less readable imo.

But looks like there's no other choice. We could keep it like you did for now.
Also I would have to be careful to not accidently use template specializations in classes anymore in the future.

AdrienTD commented on GuiUtils.cpp:
```diff
@@ -1,10 +1,9 @@
 #include "GuiUtils.h"
 #include <filesystem>
 #define WIN32_LEAN_AND_MEAN
-#define NOMINMAX
```
Is there a reason for removing NOMINMAX from every file and putting as global-wide define instead?
Might be simplier to just add NOMINMAX to files that were missing it, and leave the project file as is.

madebr replied:
The MSVC compiler defining `min` and `max` macros is the weird duckling.
It even conflicts with `std::numeric_limits<T>::max`.
So I assumed making it disappear globally so you don't need to think about it is the nicest approach (also less duplication)

AdrienTD replied:
So far I've always put NOMINMAX in the cpp, and I wouldn't put the define globally in the project file.
If a cpp needs to use platform-specific headers, it is responsible to import them properly.

Not every cpp needs to have NOMINMAX, and whenever you want to compile the file manually from command line or with a different build system (CMake etc.) you would also have to remember to define it.

Ideally most cpp files should be platform-independent, and very few use Windows APIs, so I don't think duplicating it is an issue.

AdrienTD commented on ModelImporter.h:
```diff
@@ -1,13 +1,11 @@
 struct Mesh;
 struct Chunk;
 
-namespace std::filesystem {
-	class path;
-}
```
What error did this cause?
It can be benefical to have std::filesystem::path as including <filesystem> can decrease build time a bit.

(Now it's true that whenever a header using std::filesystem::path is included in a cpp, good chance that std::filesystem::path is going to used anyway... most of the time.)

madebr replied:
[N4659 20.5.4.2.1 Namespace std](https://timsong-cpp.github.io/cppwp/n4659/namespace.std), which is the [c++ 17 draft](https://www.open-std.org/jtc1/sc22/wg21/docs/standards), says:

> The behavior of a C++ program is undefined if it adds declarations or definitions to namespace std or to a namespace within namespace std unless otherwise specified. A program may add a template specialization for any standard library template to namespace std only if the declaration depends on a user-defined type and the specialization meets the standard library requirements for the original template and is not explicitly prohibited.

This is an error that occurs due to this:
```
In file included from /usr/i686-w64-mingw32/sys-root/mingw/include/c++/filesystem:48,
                 from /home/maarten/projects/c47edit/ObjModel.cpp:5:
/usr/i686-w64-mingw32/sys-root/mingw/include/c++/bits/fs_fwd.h:318:19: error: reference to 'path' is ambiguous
  318 |   void copy(const path& __from, const path& __to, copy_options __options);
      |                   ^~~~
/usr/i686-w64-mingw32/sys-root/mingw/include/c++/bits/fs_fwd.h:58:9: note: candidates are: 'class std::filesystem::__cxx11::path'
   58 |   class path;
      |         ^~~~
In file included from /home/maarten/projects/c47edit/ObjModel.cpp:1:
/home/maarten/programming/c47edit/ObjModel.h:11:15: note:                 'class std::filesystem::path'
   11 |         class path;
      |               ^~~~
```

madebr replied:
If you want to speed up the inclusions, I'd suggest to look into precompiled headers instead.
This will give compile time speed-ups for all compile units.

AdrienTD replied:
You're right, then indeed we can just include <filesystem> in the header.

Aztec2012 commented:
AppVeyor build failed

madebr commented:
```
error C2338: static_assert failed: 'Unicode support requires compiling with /utf-8'
```
Since it is happening in libfmt, I suppose this has nothing to do with this pr but with the dependency.
I'll apply [a patch similar to this one to the vcxproj](https://github.com/python/cpython/commit/68d6bc798b34eccabdfbf94e563273759c4cef1f)

madebr committed:
libfmt assert needs /utf-8 for unicode support

This fixes the following failure:
C:\Tools\vcpkg\installed\x64-windows\include\fmt\base.h(458,28): error C2338: static_assert failed: 'Unicode support requires compiling with /utf-8' [C:\projects\c47edit\c47edit.vcxproj]

AdrienTD reviewed (commented):
Oops I just realised I had to submit the review.
I do have some questions about some of the changes.

madebr committed:
Add missing includes for memcpy

madebr commented:
With these 2 last commits, the only error remaining for targeting Linux are the `windows.h` includes (and its use of the win api).

If https://github.com/0danny/re47 ever gets somewhere, it should be easy to port everything to a platform agnostic toolkit, such as SDL.

AdrienTD reviewed (changes requested):
The commit **"Avoid windows-only _wfopen_s" 89e34e1** needs to be reverted, as this breaks the code.

On Windows, fopen expects the string to be in ANSI which is limited to 256 characters depending on the system locale.

If the filename happens to contain characters that are not part of ANSI, they will be replaced with a ? character when converting the string, and it will fail to open the file.

madebr commented:
> The commit **"Avoid windows-only _wfopen_s" [89e34e1](https://github.com/AdrienTD/c47edit/commit/89e34e15a4c590cfa898658d1911cc68a70c8b6f)** needs to be reverted, as this breaks the code.

I removed it from the patch set.

AdrienTD committed:
Merge branch 'master' into mingw-fixes
